### PR TITLE
Bitsandbytes precision plugin

### LINKED ISSUE #18295: lightning usage with bitsandbytes
### Description & Motivation

currently lightning allows training with fp16 and bf16/bf16-mixed. can these settings be used for training with bitandbytes' NF4 precision? 

### Pitch

if it's not possible to use lightning for NF4 finetuning, it would be nice to extend it to be able to allow using nf4 with Trainer. 

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #18295
Closes #18559

### CODE REVIEW

File: tests/tests_fabric/test_connector.py [hunk 1]
```diff
@@ -1088,95 +1086,3 @@ def test_xla_fsdp_automatic_strategy_selection(monkeypatch, tpu_available):
 
     if added_fsdp:
         strategies.STRATEGY_REGISTRY.pop("fsdp")
-
-
-def test_connector_transformer_engine(monkeypatch):
```
Comment (carmocca) [inline tests/tests_fabric/test_connector.py:1093]: This test is moved to its own file

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py [hunk 2]
```diff
@@ -1064,22 +1062,3 @@ def test_connector_num_nodes_input_validation():
 def test_precision_selection(precision_str, strategy_str, expected_precision_cls):
     connector = _AcceleratorConnector(precision=precision_str, strategy=strategy_str)
     assert isinstance(connector.precision_plugin, expected_precision_cls)
-
-
-def test_connector_transformer_engine(monkeypatch):
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:1069]: This test is moved to its own file

File: docs/source-pytorch/common/precision_intermediate.rst [hunk 3]
```diff
@@ -158,18 +161,71 @@ Under the hood, we use `transformer_engine.pytorch.fp8_autocast <https://docs.nv
...
+* nf4-dq: "dq" stands for "Double Quantization" which reduces the average memory footprint by quantizing the quantization constants. In average, this amounts to about 0.37 bits per parameter (approximately 3 GB for a 65B model).
+* fp4: Uses regular float 4-bit data type.
+* fp4-dq: "dq" stands for "Double Quantization" which reduces the average memory footprint by quantizing the quantization constants. In average, this amounts to about 0.37 bits per parameter (approximately 3 GB for a 65B model).
+* int8: Uses unsigned int8 data type.
+* int8-training: Meant for int8 activations with fp16 precision weights.
+
+While these techniques store weights in 4 or 8 bit, the computation still happens in 16 or 32-bit (float16, bfloat16, float32).
+This is configurable via the dtype argument in the plugin.
+
+Quantizing the model will dramatically reduce the weight's memory requirements but might impact performance negatively.
```
Comment (awaelchli) [inline docs/source-pytorch/common/precision_intermediate.rst:182]: Here, what you mean is performance in terms of predictive performance (e.g. accuracy of the classifier), not runtime performance. This could be clarified.

Comment (awaelchli) [inline docs/source-pytorch/common/precision_intermediate.rst:182]:
```suggestion
Quantizing the model will dramatically reduce the weight's memory requirements but might impact predictive performance negatively.
```

Comment (awaelchli) [inline docs/source-pytorch/common/precision_intermediate.rst:182]:
```suggestion
Quantizing the model will dramatically reduce the weight's memory requirements but might impact the accuracy of your model negatively.
```

File: docs/source-pytorch/common/precision_intermediate.rst [hunk 4]
```diff
@@ -158,18 +161,71 @@ Under the hood, we use `transformer_engine.pytorch.fp8_autocast <https://docs.nv
...
+    from lightning.pytorch.plugins import BitsandbytesPrecisionPlugin
+
+    # this will pick out the compute dtype automatically, by default `bfloat16`
+    precision = BitsandbytesPrecisionPlugin("nf4-dq")
+    trainer = Trainer(plugins=precision)
 
-It is possible to further reduce the memory usage of the optimizer states by using third-party libraries like `bitsandbytes <https://github.com/TimDettmers/bitsandbytes>`_.
-You can configure it in your LightningModule by overriding ``configure_optimizers``.
+    # Customize the dtype, or skip some modules
+    precision = BitsandbytesPrecisionPlugin("int8-training", dtype=torch.float16, skips={"lm_head"})
```
Comment (awaelchli) [inline docs/source-pytorch/common/precision_intermediate.rst:195]: Typically arguments that describe an action are in the imperative form. "skips" could both be a verb or a noun. A better choice might be ignore or exclude. Options that I like more:

- ignore
- exclude
- ignore_modules 
- exclude _modules
- skip
- skip_modules

The `_modules` version would make it very clear we are expecting module names not parameter names.

File: docs/source-pytorch/common/precision_intermediate.rst [hunk 5]
```diff
@@ -158,18 +161,71 @@ Under the hood, we use `transformer_engine.pytorch.fp8_autocast <https://docs.nv
...
+            self.model = MyModel()
+
+
+.. note::
+
+    Only supports CUDA devices and the Linux operating system. Windows users should use
+    `WSL2 <https://learn.microsoft.com/en-us/windows/ai/directml/gpu-cuda-in-wsl>`__.
+
+
+This plugin does not take care of replacing your optimizer with an 8-bit optimizer e.g. `bitsandbytes.optim.Adam8bit``.
```
Comment (awaelchli) [inline docs/source-pytorch/common/precision_intermediate.rst:211]:
```suggestion
This plugin does not take care of replacing your optimizer with an 8-bit optimizer e.g. ``bitsandbytes.optim.Adam8bit``.
```
unbalanced ticks

File: docs/source-fabric/fundamentals/precision.rst [hunk 6]
```diff
@@ -207,6 +210,73 @@ See also: :doc:`../advanced/model_init`
 ----
 
 
+*****************************
```
Comment (awaelchli) [inline docs/source-fabric/fundamentals/precision.rst:213]: Same suggestions as in the the pytorch-docs.

File: src/lightning/fabric/plugins/precision/bitsandbytes.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        if not any(isinstance(m, torch.nn.Linear) for m in module.modules()):
+            raise TypeError(
+                "You are using the bitsandbytes precision plugin, but your model has no Linear layers. This plugin"
+                " won't work for your model."
+            )
+
+        # convert modules if they haven't been converted already
+        if not any(
+            "bitsandbytes.nn" in m.__module__ or "precision.bitsandbytes" in m.__module__ for m in module.modules()
+        ):
```
Comment (awaelchli) [inline src/lightning/fabric/plugins/precision/bitsandbytes.py:109]: Why couldn't we check for a subclass?

File: src/lightning/fabric/plugins/precision/bitsandbytes.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    warnings.filterwarnings(
+        "ignore", message=r"MatMul8bitLt: inputs will be cast from .* to float16 during quantization"
+    )
+    import bitsandbytes
+
+    return bitsandbytes
+
+
+if _BITSANDBYTES_AVAILABLE:
+    bnb = _import_bitsandbytes()
```
Comment (awaelchli) [inline src/lightning/fabric/plugins/precision/bitsandbytes.py:156]: If bitsandbytes is installed but not used by the user, this code would still run at import time of lightning. Could we avoid this? The idea is that

- the two utility functions below don't need to be guarded
- the layer definitions could be moved to a function and be returned as a dict, from which you then choose the one you need

File: tests/tests_fabric/plugins/precision/test_bitsandbytes.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    module._Int8LinearInference = ModuleMock
+    module._FP4DQLinear = ModuleMock
+    module._NF4DQLinear = ModuleMock
+    module._Linear4bit = ModuleMock
+    module._Linear8bitLt = ModuleMock
+
+    precision = BitsandbytesPrecision("nf4", dtype=torch.float16)
+    connector = _Connector(plugins=precision)
+    assert connector.precision is precision
+    assert precision.dtype == torch.float16
```
Comment (awaelchli) [inline tests/tests_fabric/plugins/precision/test_bitsandbytes.py:53]: could we move the connector-specific unit tests to test_connector?

File: tests/tests_fabric/plugins/precision/test_transformer_engine.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    monkeypatch.setitem(sys.modules, "transformer_engine", transformer_engine_mock)
+    monkeypatch.setitem(sys.modules, "transformer_engine.pytorch", Mock())
+    recipe_mock = Mock()
+    monkeypatch.setitem(sys.modules, "transformer_engine.common.recipe", recipe_mock)
+
+    connector = _Connector(precision="transformer-engine")
+    assert isinstance(connector.precision, TransformerEnginePrecision)
+    assert connector.precision.dtype is torch.bfloat16
+    connector = _Connector(precision="transformer-engine-float16")
+    assert connector.precision.dtype is torch.float16
```
Comment (awaelchli) [inline tests/tests_fabric/plugins/precision/test_transformer_engine.py:40]: same suggestion

File: tests/tests_fabric/test_connector.py (hunk 1 above)
Comment (awaelchli) [reply tests/tests_fabric/test_connector.py:1093]: My suggestion is to have the connector assertions here, while doing plugin specific unit tests in the respective file

**Review by awaelchli — COMMENTED** [review]
🎉

File: src/lightning/fabric/plugins/precision/bitsandbytes.py (hunk 8 above)
Comment (carmocca) [reply src/lightning/fabric/plugins/precision/bitsandbytes.py:156]: > the layer definitions could be moved to a function and be returned as a dict, from which you then choose the one you need

That is a strange pattern. How does this work under the hood? Does Python know that the class definition is the same across calls? Or would each call create new instances with the same name?

To be clear, you are suggesting?
```
def bnb_linears():
    import bitsandbytes as bnb

	class _Linear4bit(bnb.nn.Linear4bit):
	    ...

    ...

	return {"fp4": _Linear4bit, ...}
```

File: tests/tests_fabric/plugins/precision/test_bitsandbytes.py (hunk 9 above)
Comment (carmocca) [reply tests/tests_fabric/plugins/precision/test_bitsandbytes.py:55]: This is a bit annoying to do because there's like 30 lines of mocking setup which would now be dupe in two tests. Knowing this, do you prefer to split it?

File: src/lightning/fabric/plugins/precision/bitsandbytes.py (hunk 8 above)
Comment (awaelchli) [reply src/lightning/fabric/plugins/precision/bitsandbytes.py]: No I guess it's not good to return new class objects everytime the function is called. All I'm saying is we should try to make the definition lazy. Maybe with `__getattr__` defined on the module. 

Not a blocker for this PR but we should try it, even in a follow up is fine with me. 
Because we've seen in the past how importing these 3rd party libraries ahead of time can have unintended side effects or slowdowns for users who don't even use these features. Examples from the past are deepspeed, torchelastic, logger integration libraries, torchtext etc.

File: tests/tests_fabric/plugins/precision/test_bitsandbytes.py (hunk 9 above)
Comment (awaelchli) [reply tests/tests_fabric/plugins/precision/test_bitsandbytes.py:55]: Fine with me if it's more practical this way. Could also think about a shared fixture that mocks if that makes it more managable.

**Review by awaelchli — APPROVED** [review]
Great feature. We can start testing it now with lit-gpt. 
@carmocca Would you like to also open an issue about the design concern we had around composing plugins in the future?

File: tests/tests_fabric/plugins/precision/test_bitsandbytes.py (hunk 9 above)
Comment (carmocca) [reply tests/tests_fabric/plugins/precision/test_bitsandbytes.py:55]: Maybe in the future if we add more tests

File: src/lightning/fabric/plugins/precision/bitsandbytes.py (hunk 8 above)
Comment (carmocca) [reply src/lightning/fabric/plugins/precision/bitsandbytes.py]: Let me try __getattr__

**Comment by carmocca** [comment]
@awaelchli Opened https://github.com/Lightning-AI/lightning/issues/18679. Also working to update lit-gpt in https://github.com/Lightning-AI/lit-gpt/pull/596

**Review by Borda — APPROVED** [review]
